cayenne-di/src/main/java/org/apache/cayenne/di/spi/DIGraph.java

```diff
@@ -68,6 +68,7 @@ public void addAll(Collection<V> vertexes) {
 	public void add(V from, V to) {
 		this.add(from);
 		this.add(to);
+		neighbors.get(to).remove(from);
```

Comment:
This change breaks Module providers circular dependency check, see `ModuleLoader.loadModules()`. An easy fix could be just add option to the `DIGraph` use different logic in List creation and in modules loading.

Reply:
Added test case for that: https://github.com/apache/cayenne/blob/master/cayenne-di/src/test/java/org/apache/cayenne/di/spi/ModuleLoaderTest.java#L64